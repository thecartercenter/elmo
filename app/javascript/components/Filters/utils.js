import isEmpty from 'lodash/isEmpty';
import mapKeys from 'lodash/mapKeys';
import queryString from 'query-string';

import { SUBMITTER_TYPES } from './SubmitterFilter/utils';

const MAX_HINTS_BEFORE_ELLIPSIZE = 1;

/**
 * Stringified controller_name from Rails.
 */
export const CONTROLLER_NAME = {
  RESPONSES: '"responses"',
};

/**
 * Returns a new instance of FiltersModel.
 *
 * Generally this should be added to a top-level Provider and only used once.
 */
export function provideFiltersStore(FiltersModel, initialState) {
  const store = new FiltersModel(initialState);

  if (process.env.NODE_ENV === 'development') {
    // Debug helper.
    window.store = store;
  }

  return store;
}

/**
 * Given a list of hints (e.g. currently selected form names for the form filter button),
 * stringify them to be displayed on the button itself.
 */
export function getButtonHintString(hints) {
  if (isEmpty(hints)) {
    return '';
  }

  const joinedHints = hints.length > MAX_HINTS_BEFORE_ELLIPSIZE
    ? `${hints.length} filters`
    : hints.join(', ');

  return ` (${joinedHints})`;
}

/**
 * Given an item ID, find it in the list of items and return its `name`.
 */
export function getItemNameFromId(allItems, searchId, nameKey = 'name') {
  const item = allItems.find(({ id }) => searchId === id);
  return (item && item[nameKey]) || 'Unknown';
}

/**
 * Given a leftQingId, find it in the list of all questions and return the name.
 */
export function getQuestionNameFromId(allQuestions, searchId) {
  return getItemNameFromId(allQuestions, searchId, 'code');
}

/**
 * Converts a list of data from the backend into something Select2 understands, e.g.
 * [{ id: '1', name: 'One' }, ...] => [{ id: '1', text: 'One' }, ...]
 */
export function parseListForSelect2(allItems) {
  return allItems.map((item) =>
    mapKeys(item, (value, key) => (key === 'name' ? 'text' : key)));
}

/**
 * Given all of the different filter states,
 * return a stringified version for the backend.
 */
export function getFilterString({
  selectedFormIds,
  conditionSetStore,
  isReviewed,
  selectedSubmittersForType,
  advancedSearchText,
  startDate,
  endDate,
}) {
  const allQuestions = conditionSetStore.refableQings;
  const questionFilters = conditionSetStore.conditions
    .filter(({ leftQingId, currTextValue, remove }) => leftQingId && currTextValue && !remove)
    .map(({ leftQingId, currTextValue }) =>
      `{${getQuestionNameFromId(allQuestions, leftQingId)}}:${JSON.stringify(currTextValue)}`);

  const submitterParts = SUBMITTER_TYPES.map((type) => {
    const selectedSubmitterIds = selectedSubmittersForType[type].map(({ id }) => id);
    return isEmpty(selectedSubmitterIds) ? null : `${type}-id:(${selectedSubmitterIds.join('|')})`;
  });

  const dateParts = [];
  if (startDate) {
    dateParts.push(`submit-date>=${startDate.format('YYYY-MM-DD')}`);
  }
  if (endDate) {
    dateParts.push(`submit-date<=${endDate.format('YYYY-MM-DD')}`);
  }

  const parts = [
    isEmpty(selectedFormIds) ? null : `form-id:(${selectedFormIds.join('|')})`,
    ...questionFilters,
    isReviewed == null ? null : `reviewed:${isReviewed ? '1' : '0'}`,
    ...submitterParts,
    dateParts.join(' '),
    advancedSearchText,
  ].filter(Boolean);

  return parts.join(' ');
}

/**
 * Reload the page with the given search,
 * resetting the page number.
 */
export function submitSearch(filterString) {
  const parsed = queryString.parse(window.location.search);
  // Params will be removed from the URL if `undefined`.
  const search = filterString || undefined;
  const page = undefined;
  const params = queryString.stringify({ ...parsed, search, page });

  window.location.assign(params
    ? `?${params}`
    : window.location.pathname);
}

/**
 * Returns true if the given param name exists and is non-empty.
 */
export function isQueryParamTruthy(paramName) {
  const parsed = queryString.parse(window.location.search);
  return Boolean(parsed[paramName]);
}
