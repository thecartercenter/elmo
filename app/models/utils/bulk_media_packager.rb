# frozen_string_literal: true

require "sys/filesystem"
require "zip"
require "fileutils"

module Utils
  # Utility to support a bulk image download operation
  class BulkMediaPackager
    include ActiveModel::Model
    # Space we want to leave on disk in mib
    DISK_ALLOWANCE = 2048
    TMP_DIR = "tmp/bulk_media"

    attr_accessor :ability, :search, :operation

    def media_objects_scope
      responses = Response.accessible_by(@ability, :export)
      responses = apply_search_scope(responses, @search, @operation.mission) if @search.present?
      responses.joins(:answers)
        .joins("INNER JOIN media_objects ON media_objects.answer_id = answers.id")
        .joins("INNER JOIN active_storage_attachments ON
          active_storage_attachments.record_id = media_objects.id")
        .joins("INNER JOIN active_storage_blobs ON
          active_storage_blobs.id = active_storage_attachments.blob_id")
    end

    def calculate_media_size
      media_objects_scope.sum("active_storage_blobs.byte_size")
    end

    def space_on_disk?
      stat = Sys::Filesystem.stat("/")
      # need to leave space for images, zip file, and copy of zip file while attaching to operation
      space_left = bytes_to_mib(stat.block_size * stat.blocks_available) -
        bytes_to_mib(calculate_media_size * 2)
      space_left >= DISK_ALLOWANCE
    end

    def download_and_zip_images
      FileUtils.mkdir_p(Rails.root.join(TMP_DIR))
      media_ids = media_objects_scope.pluck("media_objects.id")
      filename = "#{@operation.mission.compact_name}-media-#{Time.current.to_s(:filename_datetime)}.zip"
      zipfile_name = Rails.root.join(TMP_DIR, filename)

      Zip::File.open(zipfile_name, Zip::File::CREATE) do |zipfile|
        media_ids.each do |media_id|
          zip_media(::Media::Object.find(media_id).item, zipfile)
        rescue Zip::EntryExistsError
          next
        end
      end
      zipfile_name
    end

    private

    def zip_media(attachment, zipfile)
      attachment.open do |file|
        filename = attachment.filename.to_s
        new_path = Rails.root.join(TMP_DIR, filename)
        FileUtils.mv(file.path, new_path)
        zip_entry = Utils::ZipEntry.new(new_path, filename)
        zipfile.add(zip_entry, new_path)
      end
    end

    def apply_search_scope(responses, search, mission)
      ResponsesSearcher.new(relation: responses, query: search, scope: {mission: mission}).apply
    end

    def bytes_to_mib(bytes)
      bytes / 1024 / 1024
    end
  end
end
