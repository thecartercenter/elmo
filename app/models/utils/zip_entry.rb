# frozen_string_literal: true

require "sys/filesystem"
require "zip"
require "fileutils"

module Utils
  class ZipEntry < ::Zip::Entry

    def clean_up
      FileUtils.rm @zipfile
    end

  end
end
