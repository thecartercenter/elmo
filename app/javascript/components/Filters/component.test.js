import pick from 'lodash/pick';
import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'mobx-react';

import { quietMount } from '../../testUtils';
import { getFiltersStore } from './testUtils';

import { CONTROLLER_NAME } from './utils';
import Component from './component';
import { SUBMITTER_TYPES, submitterType } from './SubmitterFilter/utils';

jest.mock('../conditions/ConditionSetFormField/component', () => 'ConditionSetFormField');
jest.mock('../conditions/AddConditionLink/component', () => 'AddConditionLink');

let defaultProps = {};
resetDefaultProps();

function resetDefaultProps() {
  const filtersStore = getFiltersStore();

  defaultProps = {
    filtersStore,
    conditionSetStore: filtersStore.conditionSetStore,

    // Initial props.
    ...pick(filtersStore, ['allForms', 'selectedFormIds', 'isReviewed', 'advancedSearchText']),
    allUsers: filtersStore.allSubmittersForType[submitterType.USER],
    selectedUsers: filtersStore.selectedSubmittersForType[submitterType.USER],
    allGroups: filtersStore.allSubmittersForType[submitterType.GROUP],
    selectedGroups: filtersStore.selectedSubmittersForType[submitterType.GROUP],
    controllerName: CONTROLLER_NAME.RESPONSES,
  };
}

it('renders as expected (responses page)', () => {
  const wrapper = shallow(<Component {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

it('renders as expected (other page)', () => {
  const wrapper = shallow(
    <Component
      {...defaultProps}
      controllerName="foo"
    />,
  );
  expect(wrapper).toMatchSnapshot();
});

describe('integration', () => {
  // Re-import Filters with unmocked @inject decorator so we can deep render.
  jest.resetModules();
  jest.unmock('mobx-react');
  // eslint-disable-next-line no-shadow
  const Component = require('./component').default;

  let wrapper;

  beforeEach(() => {
    window.location.assign.mockClear();

    resetDefaultProps();

    wrapper = quietMount(
      <Provider
        filtersStore={defaultProps.filtersStore}
        conditionSetStore={defaultProps.filtersStore.conditionSetStore}
      >
        <Component {...defaultProps} />
      </Provider>,
    );
  });

  it('navigates on apply form filter', () => {
    wrapper.find('Button#form-filter').simulate('click');
    const overlay = quietMount(wrapper.find('OverlayTrigger#form-filter').prop('overlay'));
    // Call prop directly since Select2 is stubbed out.
    overlay.find('Select2').prop('onSelect')({ target: { value: defaultProps.filtersStore.allForms[0].id } });

    expectLocationToChangeOnClick(overlay.find('Button.btn-apply'));
  });

  it('navigates on apply question filter', () => {
    wrapper.find('Button#question-filter').simulate('click');
    const overlay = quietMount(wrapper.find('OverlayTrigger#question-filter').prop('overlay'));

    expectLocationToChangeOnClick(overlay.find('Button.btn-apply'));
  });

  it('navigates on apply reviewed filter', () => {
    wrapper.find('Button#reviewed-filter').simulate('click');
    const overlay = quietMount(wrapper.find('OverlayTrigger#reviewed-filter').prop('overlay'));
    overlay.find('#no').simulate('click');

    expectLocationToChangeOnClick(overlay.find('Button.btn-apply'));
  });

  it('navigates on apply submitter filter', () => {
    wrapper.find('Button#submitter-filter').simulate('click');
    const overlay = quietMount(wrapper.find('OverlayTrigger#submitter-filter').prop('overlay'));

    SUBMITTER_TYPES.forEach((type) => {
      const { id, name } = defaultProps.filtersStore.allSubmittersForType[type][0];
      // Call prop directly since Select2 is stubbed out.
      overlay.find(`Select2#${type}`).prop('onSelect')({ params: { data: { id, text: name } } });
    });

    expectLocationToChangeOnClick(overlay.find('Button.btn-apply'));
  });

  it('navigates on apply advanced search', () => {
    wrapper.find('.search-str').simulate('change', { target: { value: 'something else' } });
    expectLocationToChangeOnClick(wrapper.find('Button.btn-advanced-search'));
  });
});

function expectLocationToChangeOnClick(element) {
  expect(window.location.assign.mock.calls.length).toBe(0);
  element.simulate('click');
  expect(window.location.assign.mock.calls.length).toBe(1);
  expect(window.location.assign.mock.calls[0]).toMatchSnapshot();
}
