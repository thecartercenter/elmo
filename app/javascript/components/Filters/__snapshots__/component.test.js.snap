// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration navigates on apply advanced search 1`] = `
Array [
  "?search=form-id%3A%282%29%20%7BOne%7D%3A%225%22%20reviewed%3A1%20submitter-id%3A%28B%29%20group-id%3A%28B%29%20something%20else",
]
`;

exports[`integration navigates on apply form filter 1`] = `
Array [
  "?search=form-id%3A%281%29%20reviewed%3A1%20submitter-id%3A%28B%29%20group-id%3A%28B%29%20query",
]
`;

exports[`integration navigates on apply question filter 1`] = `
Array [
  "?search=form-id%3A%282%29%20%7BOne%7D%3A%225%22%20reviewed%3A1%20submitter-id%3A%28B%29%20group-id%3A%28B%29%20query",
]
`;

exports[`integration navigates on apply reviewed filter 1`] = `
Array [
  "?search=form-id%3A%282%29%20%7BOne%7D%3A%225%22%20reviewed%3A0%20submitter-id%3A%28B%29%20group-id%3A%28B%29%20query",
]
`;

exports[`integration navigates on apply submitter filter 1`] = `
Array [
  "?search=form-id%3A%282%29%20%7BOne%7D%3A%225%22%20reviewed%3A1%20submitter-id%3A%28A%29%20group-id%3A%28A%29%20query",
]
`;

exports[`renders as expected (other page) 1`] = `
<div
  className="filters"
>
  <AdvancedSearchFilter
    onClear={[Function]}
    onSubmit={[Function]}
    renderInfoButton={false}
  />
</div>
`;

exports[`renders as expected (responses page) 1`] = `
<div
  className="filters"
>
  <ButtonToolbar>
    <i
      className="fa fa-filter"
    />
    <FormFilter
      onSubmit={[Function]}
    />
    <QuestionFilter
      onSubmit={[Function]}
    />
    <FormFilter
      onSubmit={[Function]}
    />
    <SubmitterFilter
      onSubmit={[Function]}
    />
    <DateFilter
      onSubmit={[Function]}
    />
  </ButtonToolbar>
  <AdvancedSearchFilter
    onClear={[Function]}
    onSubmit={[Function]}
    renderInfoButton={true}
  />
</div>
`;
